packages/Ecotone/src/Messaging/Config/EnterpriseLicenceDecider.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Ecotone\Messaging\Config;
+
+use Ecotone\Messaging\Attribute\Enterprise;
+use Ecotone\Messaging\Config\Container\Definition;
+use Ecotone\Messaging\Config\Container\Reference;
+use Ecotone\Messaging\Handler\ClassDefinition;
+use Ecotone\Messaging\Handler\InterfaceToCall;
+use Ecotone\Messaging\Handler\TypeDescriptor;
+
+/**
+ * licence Enterprise
+ *
+ * Enterprise features are licenced under separate Enterprise licence.
+ * Therefore to make us of those features you've to obtain a licence.
+ * For more details contact us at: sales@simplycodedsoftware.com
+ */
+final class EnterpriseLicenceDecider
+{
+    public function __construct(private bool $isOnEnterpriseLicence)
+    {
+
+    }
+
+    public function hasEnterpriseLicence(): bool
+    {
+        return $this->isOnEnterpriseLicence;
+    }
+
+    public function isEnabledSpecificallyFor(InterfaceToCall|ClassDefinition $interfaceToCall): bool
+    {
+        if (!$this->hasEnterpriseLicence()) {
+            return false;
+        }
+
+        $type = TypeDescriptor::create(Enterprise::class);
+        if ($interfaceToCall instanceof ClassDefinition) {
+            return $interfaceToCall->hasClassAnnotation(TypeDescriptor::create(Enterprise::class));
+        }
+
+        return $interfaceToCall->hasAnnotation($type);
+    }
+
+    public static function prepareDefinition(string $className, Reference|Definition $openCoreService, Reference|Definition $enterpriseService): Definition
+    {
+        return new Definition(
+            $className,
+            [
+                Reference::to(self::class),
+                $openCoreService,
+                $enterpriseService
+            ],
+            factory: [self::class, "create"]
+        );
+    }
+
+    public static function create(self $enterpriseModeDecider, object $openCoreService, object $enterpriseService): object
+    {
+        if ($enterpriseModeDecider->hasEnterpriseLicence()) {
+            return $enterpriseService;
+        }
+
+        return $openCoreService;
+    }
```

Comment:
It should work like that :

```suggestion
    public static function prepareDefinition(string $className, Reference|Definition $openCoreService, Reference|Definition $enterpriseService): Definition
    {
        return new Definition(
            $className,
            [
                $openCoreService,
                $enterpriseService
            ],
            factory: [self::class, "decide"]
        );
    }

    public function decide(object $openCoreService, object $enterpriseService): object
    {
        if ($this->hasEnterpriseLicence()) {
            return $enterpriseService;
        }

        return $openCoreService;
    }
```

Reply:
Can you explain the difference? I am having hard time spotting that. 

> factory: [self::class, "decide"]

Shouldn't the factory method be static?

Reply:
The factory can be an object instance, it depends if the method is static or not. It is similar to symfony : https://symfony.com/doc/current/service_container/factories.html#non-static-factories

Maybe I am wrong with my suggestion and you should pass a reference (not just the class name) as first array key:
```php
        return new Definition(
            $className,
            [
                $openCoreService,
                $enterpriseService
            ],
            factory: [Reference::to(self::class), "decide"]
        );
```

Reply:
I will cover conditional registering of Services with next Enterprise related PR

Reply:
So [in here](https://github.com/ecotoneframework/ecotone-dev/pull/342/files#diff-52504a6b952169e62f6e218430717248d53a8ed55d1cff3786d4e790438457caR330), being able to differentiate in what mode we currently are.